Fixes (mainly build system) for Trilinos promotion (#2511)
kokkos/kokkos: scripts/trilinos-integration/README.md

![Kokkos](https://avatars2.githubusercontent.com/u/10199860?s=200&v=4)

# Kokkos: Promotion Test Debugging

This explains the use (and basic implementation details) of promotion testing
new Kokkos and Kokkos Kernels branches. We first introduce a test for first

> validating a "clean" Trilinos branch that all tests should be passing on.

Reviewer: Not sure what you mean by "Trilinos branch".

Author: Tried to add clarifying text

Once validated on a clean promotion, we show how to test the same Trilinos branch
with an updated Kokkoks and Kokkos Kernels.

# Clean Promotion Test

> There is a script called `clean_promotion_test` that tests a self-contained

Reviewer: Say the script can be found in `scripts/trilinos-integration/`.  My first reaction was where can I find it.

Author: Done.

Trilinos branch with its default included Kokkos and Kokkos Kernels packages.

> The script takes two arguments:

Reviewer 2: It appears that the script assumes that you are in the [whatever]/kokkos/scripts/trilinos-integration directory, but that is not explicitly stated.  I think it might be better to not have such hard bindings in the scripts...The biggest issue / only thing I had to tweak calling it from somewhere else is the path to nvcc_wrapper.

````
./clean_promotion_test <ENV_FILE> <TRILINOS_ROOT>
````
For running CUDA tests on the platform called White, e.g.
````
./clean_promotion_test white_cuda_env.sh $TRILINOS_ROOT
````
Based on the Trilinos path and ENV file, a unique hash `X` is generated.

> A CMake configuration is then run in the folder `trilinos-promotion-X`.

Reviewer 2: when I tested this, I got a directory called promotion-test-X

Author: Yes, you did. I fixed the docs. I also changed the "clean" tests to run in `clean-test-X`

Output of the configuration will appear in a `config.out`.

> If the configuration succeeds, you can `cd` in to the folder and run:

Reviewer 3: ```suggestion
If the configuration succeeds, you can `cd` into the folder and run:
```

````
> source ../<ENV_FILE>
> make -j
````
to validate the build. The configure and build steps are deliberately separated
to allow incremental debugging of each step.

# New Branch Promotion Test
If the clean test is passing, you can now test your updates to Kokkos.
To start the tests, there is a script that now takes four arguments:
````
./config_promotion_test <ENV_FILE> <TRILINOS_ROOT> <KOKKOS_ROOT> <KERNELS_ROOT>
````
where the two additional arguments are the locations of Kokkos and Kokkos Kernels
branches containing updated. Again, a unique hash `X` is generated.
A CMake configuration is then run in the folder `trilinos-promotion-X`.

> If the configuration succeeds, you can `cd` in to the folder and run:

Reviewer 3: ```suggestion
If the configuration succeeds, you can `cd` into the folder and run:
```

Author: Done

> If the configuration succeeds, you can `cd` into the folder and run:

Reviewer 2: This might be assumed, but on systems like waterman and white, this is best done in on interactive node

Author: I think it should be clear how to adjust for your particular use case? If you think some extra detail is needed, though, I can add it.

Reviewer 2: This is fine...as long as we assume the target audience is versed on how to run test suites on SNL servers

````
> source ../<ENV_FILE>
> make -j
````
The script uses the the source override feature of the Trilinos build system.

> It creates sym links in the Trilinos folder to your updated Kokkos

Reviewer 3: ```suggestion
It creates symlinks in the Trilinos folder to your updated Kokkos
```
